sky/admin_policy.py

```diff
@@ -58,15 +75,64 @@ class UserRequest:
     request_options: Optional['RequestOptions'] = None
     at_client_side: bool = False
 
+    def encode(self) -> str:
+        return _UserRequestBody(
+            task=self.task.to_yaml_config(),
+            skypilot_config=dict(self.skypilot_config),
+            request_options=self.request_options,
+            at_client_side=self.at_client_side).model_dump_json()
+
+    @classmethod
+    def decode(cls, body: str) -> 'UserRequest':
+        user_request_body = _UserRequestBody.model_validate_json(body)
+        return cls(task=sky.Task.from_yaml_config(user_request_body.task),
+                   skypilot_config=config_utils.Config.from_dict(
+                       user_request_body.skypilot_config),
+                   request_options=user_request_body.request_options,
+                   at_client_side=user_request_body.at_client_side)
+
+
+class _MutatedUserRequestBody(pydantic.BaseModel):
+    """Auxiliary model to validate and serialize a user request."""
+    task: Dict[str, Any]
+    skypilot_config: Dict[str, Any]
+
 
 @dataclasses.dataclass
 class MutatedUserRequest:
+    """Mutated user request."""
+
     task: 'sky.Task'
     skypilot_config: 'sky.Config'
 
+    def encode(self) -> str:
+        return _MutatedUserRequestBody(
+            task=self.task.to_yaml_config(),
+            skypilot_config=dict(self.skypilot_config)).model_dump_json()
+
+    @classmethod
+    def decode(cls, mutated_user_request_body: str) -> 'MutatedUserRequest':
+        mutated_user_request_body = _MutatedUserRequestBody.model_validate_json(
+            mutated_user_request_body)
+        return cls(task=sky.Task.from_yaml_config(
+            mutated_user_request_body.task),
+                   skypilot_config=config_utils.Config.from_dict(
+                       mutated_user_request_body.skypilot_config))
+
+
+class PolicyInterface:
+    """Interface for admin-defined policy for user requests."""
+
+    @abc.abstractmethod
+    def apply(self, user_request: UserRequest) -> MutatedUserRequest:
```

Comment:
Do we have to update the API name? It might be fine to keep it as before.

Reply:
The previous one is a class method, we should keep it as is to keep the backward-compatibility I think. Thus a new function name is needed for the abstract instance method.